Validate the existence of the model specified in --model

cdoern opened the pull request:
# Changes

the openAI client that connects to endpoint_url has a list of models the server can chat with.

We need to look through that list and ensure the user specified --model or -m exists in that list

**Which issue is resolved by this Pull Request:**
Resolves #753

russellb reviewed (changes requested):
Let's hold this until after #935 since they are so closely related.

Also - this change touches `chat`, but the same issue you're addressing here applies to `generate` as well.

russellb commented:
I removed the hold because #935 was merged. This will have to be updated to account for those changes. We should definitely validate the requested model and raise a friendly error if it's not there, but you'll have to account for the code from #935 that handles the case of running a non-default model from `ilab serve` and then running `ilab chat` with no arguments. That has always worked even though the code asks for the wrong model name. #935 made it so the chat UI reflected the actual model we connected to.

When something specific and non-default is requested though, whether it's local or a remote endpoint URL, this validation is good and can help us give a more friendly error than getting a 404 Not Found error a bit later.

nathan-weinberg commented:
@Mergifyio rebase

leseb commented on src/instructlab/chat/chat.py:
```diff
@@ -478,6 +478,14 @@ def chat_cli(
+            modelid_found = True
+
+    if not modelid_found:
+        raise ChatException(f"API issue found while executing chat, Model path {model} does not exist on the server.")
```
```suggestion
        raise ChatException(f"Model path {model} does not exist on the server.")
```

Also, I'd print the list of available models to help the user further. Thanks!

hickeyma replied:
Might read better as: "Model {model} is not served by the server."?

leseb reviewed (commented):
I believe this is getting close, thanks for your patience.

hickeyma reviewed (changes requested):
Nice check, thanks for adding @cdoern. One small nit.

russellb reviewed (changes requested):
I think this comment still applies?  let me know if I missed something

> Also - this change touches chat, but the same issue you're addressing here applies to generate as well.

cdoern commented:
sorry @russellb I totally missed that! Fixed now, want to let the tests run since I realized running `ilab chat` fails since the default for the model flag is now incorrect

russellb reviewed (approved):
lgtm pending CI

russellb reviewed (changes requested):
spoke too soon, CI failed

>   Executing chat failed with: Model merlinite-7b-lab-Q4_K_M is not served by the server.

cdoern commented:
@russellb I did the following and it worked:

1. nano config.yaml, change it to: 
```chat:
  context: default
  greedy_mode: false
  logs_dir: data/chatlogs
  max_tokens: null
  model: models/merlinite-7b-lab-Q4_K_M.gguf
  session: null
  vi_mode: false
  visible_overflow: true
general:
  log_level: INFO
generate:
  chunk_word_count: 1000
  model: models/merlinite-7b-lab-Q4_K_M.gguf
  num_cpus: 10
  num_instructions: 100
  output_dir: generated
  prompt_file: prompt.txt
  seed_file: seed_tasks.json
  taxonomy_base: origin/main
  taxonomy_path: taxonomy
serve:
  gpu_layers: -1
  host_port: 127.0.0.1:8000
  max_ctx_size: 4096
  model_path: models/granite-7b-lab-Q4_K_M.gguf
```
2. ilab serve
3. ilab chat
```
(venv) cdoern@cdoern-thinkpadp16vgen1:~/Documents/instructlab$ ilab chat
╭────────────────────────────────────────────── system ───────────────────────────────────────────────╮
│ Welcome to InstructLab Chat w/ MODELS/GRANITE-7B-LAB-Q4_K_M.GGUF (type /h for help)                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────╯
>>>                                                                                        [S][default]

```

let me know if you are doing something different

cdoern commented:
maybe you didn't specify the .gguf part?

leseb commented on scripts/functional-tests.sh:
```diff
@@ -344,31 +344,6 @@ test_model_print(){
         }
     '
 
-    # validate that we print the model from the CLI
```
Why is this being removed?

hickeyma replied:
Agree

cdoern replied:
this is testing whether the user can put something like baz.gguf, this should not work and will not. The new code only allows  -m which is in the server

cdoern replied:
`ilab chat -m bar` should really never have worked, this PR makes it so that -m only works for real, fully qualified models.

cdoern replied:
sure

hickeyma commented on src/instructlab/config.py:
```diff
@@ -19,8 +19,8 @@
 
 DEFAULT_API_KEY = "no_api_key"
 DEFAULT_CONFIG = "config.yaml"
-DEFAULT_MODEL = "merlinite-7b-lab-Q4_K_M"
-DEFAULT_MODEL_PATH = f"models/{DEFAULT_MODEL}.gguf"
```
Why are we making changes here? Is it required for the PR goal?

cdoern replied:
The default model needs to match the full path. The -m the user gives and/or the default need to be the fully qualified path to the served model, especially in vllm.

cdoern replied:
something like `merlinite-7b-lab` isn't ever going to be one of the models found when querying the server, only the full path will be recognized

hickeyma reviewed (dismissed):
@cdoern Can you explain changes that are not directly related to validating the existence of the model?

cdoern commented:
@hickeyma @russellb PTAL, I would like this fix to be in for today's release

leseb commented on src/instructlab/config.py:
```diff
@@ -19,8 +19,8 @@
 DEFAULT_CONFIG = "config.yaml"
-DEFAULT_MODEL = "merlinite-7b-lab-Q4_K_M"
-DEFAULT_MODEL_PATH = f"models/{DEFAULT_MODEL}.gguf"
+DEFAULT_MODEL = "models/merlinite-7b-lab-Q4_K_M.gguf"
+DEFAULT_MODEL_PATH = "models/merlinite-7b-lab-Q4_K_M.gguf"
```
Since both `DEFAULT_MODEL` and `DEFAULT_MODEL_PATH` share the same value. Why keep both?

cdoern replied:
Adding a todo, I think this  consolidation is out of scope for this PR as that would involve changing cmd flags since these vars auto populate those, right?

```# TODO: Consolidate --model and --model-path into one --model-path flag since we always need a path now
```

leseb replied:
I'm fine with that, another small request, could you please open an issue so we can track it more easily? Thanks!

russellb replied:
or just this and skip the issue ...

```suggestion
DEFAULT_MODEL_PATH = DEFAULT_MODEL = "models/merlinite-7b-lab-Q4_K_M.gguf"
```

leseb reviewed (commented):
Small nit question.

russellb commented:
> @russellb I did the following and it worked:
> 
> 1. nano config.yaml, change it to:
> 
> ```
>   context: default
>   greedy_mode: false
>   logs_dir: data/chatlogs
>   max_tokens: null
>   model: models/merlinite-7b-lab-Q4_K_M.gguf
>   session: null
>   vi_mode: false
>   visible_overflow: true
> general:
>   log_level: INFO
> generate:
>   chunk_word_count: 1000
>   model: models/merlinite-7b-lab-Q4_K_M.gguf
>   num_cpus: 10
>   num_instructions: 100
>   output_dir: generated
>   prompt_file: prompt.txt
>   seed_file: seed_tasks.json
>   taxonomy_base: origin/main
>   taxonomy_path: taxonomy
> serve:
>   gpu_layers: -1
>   host_port: 127.0.0.1:8000
>   max_ctx_size: 4096
>   model_path: models/granite-7b-lab-Q4_K_M.gguf
> ```
> 
> 2. ilab serve
> 3. ilab chat
> 
> ```
> (venv) cdoern@cdoern-thinkpadp16vgen1:~/Documents/instructlab$ ilab chat
> ╭────────────────────────────────────────────── system ───────────────────────────────────────────────╮
> │ Welcome to InstructLab Chat w/ MODELS/GRANITE-7B-LAB-Q4_K_M.GGUF (type /h for help)                 │
> ╰─────────────────────────────────────────────────────────────────────────────────────────────────────╯
> >>>                                                                                        [S][default]
> ```
> 
> let me know if you are doing something different

It looks like the difference is that your `config.yaml` was generated after the changes made in your PR. If you have a config using the defaults from prior to this PR, it breaks.

```
chat:
  context: default
  greedy_mode: false
  logs_dir: data/chatlogs
  max_tokens: null
  model: merlinite-7b-lab-Q4_K_M
  session: null
  vi_mode: false
  visible_overflow: true
general:
  log_level: INFO
generate:
  chunk_word_count: 1000
  model: merlinite-7b-lab-Q4_K_M
  num_cpus: 10
  num_instructions: 100
  output_dir: generated
  prompt_file: prompt.txt
  seed_file: seed_tasks.json
  taxonomy_base: origin/main
  taxonomy_path: taxonomy
serve:
  gpu_layers: -1
  host_port: 127.0.0.1:8000
  max_ctx_size: 4096
  model_path: models/merlinite-7b-lab-Q4_K_M.gguf
```

```
$ ilab serve
$ ilab chat
Executing chat failed with: Model merlinite-7b-lab-Q4_K_M is not served by the server.

cdoern commented:
> > @russellb I did the following and it worked:
> > 
> > 1. nano config.yaml, change it to:
> > 
> > ```
> >   context: default
> >   greedy_mode: false
> >   logs_dir: data/chatlogs
> >   max_tokens: null
> >   model: models/merlinite-7b-lab-Q4_K_M.gguf
> >   session: null
> >   vi_mode: false
> >   visible_overflow: true
> > general:
> >   log_level: INFO
> > generate:
> >   chunk_word_count: 1000
> >   model: models/merlinite-7b-lab-Q4_K_M.gguf
> >   num_cpus: 10
> >   num_instructions: 100
> >   output_dir: generated
> >   prompt_file: prompt.txt
> >   seed_file: seed_tasks.json
> >   taxonomy_base: origin/main
> >   taxonomy_path: taxonomy
> > serve:
> >   gpu_layers: -1
> >   host_port: 127.0.0.1:8000
> >   max_ctx_size: 4096
> >   model_path: models/granite-7b-lab-Q4_K_M.gguf
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. ilab serve
> > 3. ilab chat
> > 
> > ```
> > (venv) cdoern@cdoern-thinkpadp16vgen1:~/Documents/instructlab$ ilab chat
> > ╭────────────────────────────────────────────── system ───────────────────────────────────────────────╮
> > │ Welcome to InstructLab Chat w/ MODELS/GRANITE-7B-LAB-Q4_K_M.GGUF (type /h for help)                 │
> > ╰─────────────────────────────────────────────────────────────────────────────────────────────────────╯
> > >>>                                                                                        [S][default]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > let me know if you are doing something different
> 
> It looks like the difference is that your `config.yaml` was generated after the changes made in your PR. If you have a config using the defaults from prior to this PR, it breaks.
> 
> ```
> chat:
>   context: default
>   greedy_mode: false
>   logs_dir: data/chatlogs
>   max_tokens: null
>   model: merlinite-7b-lab-Q4_K_M
>   session: null
>   vi_mode: false
>   visible_overflow: true
> general:
>   log_level: INFO
> generate:
>   chunk_word_count: 1000
>   model: merlinite-7b-lab-Q4_K_M
>   num_cpus: 10
>   num_instructions: 100
>   output_dir: generated
>   prompt_file: prompt.txt
>   seed_file: seed_tasks.json
>   taxonomy_base: origin/main
>   taxonomy_path: taxonomy
> serve:
>   gpu_layers: -1
>   host_port: 127.0.0.1:8000
>   max_ctx_size: 4096
>   model_path: models/merlinite-7b-lab-Q4_K_M.gguf
> ```
> 
> ```
> $ ilab serve
> $ ilab chat
> Executing chat failed with: Model merlinite-7b-lab-Q4_K_M is not served by the server.
> ```

Yep, this seems expected to me since we changed the defaults. Maybe a note in the release about this?

russellb commented on src/instructlab/chat/chat.py:
```diff
@@ -482,6 +482,13 @@ def chat_cli(
     )
+    # ensure the model specified exists on the server. with backends like vllm, this is crucial.
+    if not any(model == m.id for m in client.models.list().data):
+        if not os.path.exists(model):
```
This check won't work if we're talking to a remote model endpoint. It's also making assumptions about how the model is named, which may not be a path to the model at all. That's how it works with our built-in llama-cpp, but it doesn't have to be that way.

nathan-weinberg replied:
I think one of the other points of the PR is forcing the model to be a path, since we could have duplicate model names. 

Can we even talk to a remote model endpoint at this stage? Doesn't it have to be a local file?

russellb replied:
> I think one of the other points of the PR is forcing the model to be a path, since we could have duplicate model names.
> 
> Can we even talk to a remote model endpoint at this stage? Doesn't it have to be a local file?

It does not.  You can do `ilab chat --endpoint-url https://anything.openai.compatible/v1`.

Here's a suggested alternative:

Add:

```
DEFAULT_MODEL_OLD = "merlinite-7b-lab-Q4_K_M"
```

and then:
```suggestion
        if model == DEFAULT_MODEL_OLD:
            ... friendly message about using an old config that we know is now broken with instructions on how to fix it ...
```

cdoern replied:
Ok, added this

russellb reviewed (approved):
Thanks for all the follow-up work, @cdoern.

I've played with this again with different configs and options and I'm happy with the behavior now.

cdoern commented:
@hickeyma your comments were addressed, thanks for the reviews! I added that test back and expected it to fail, I also added `DEFAULT_MODEL_OLD` which I think makes it more clear as to why those constants changed.

The most recent push incorporated @leseb's idea to print the served models when we fail. 

Dropping the blocking changes requested since all of the comments have been incorporated, and i want to ensure we have some time with this in before cutting the release :)
